### ISSUE TITLE
separate the modules GR and jlgr

### ISSUE THREAD

**Issue by heliosdrm** [jheinen/GR.jl#225]
**TL;DR**: Remove all the references to `jlgr` from the module `GR`, and export the high-level functions directly by `jlgr`.

There are some issues (e.g. #196, #207) that arise from people using low-level functions from `GR.jl` the same way as the convenience high-level functions from `jlgr`. I think that this confusion is natural, since by looking at their docstrings some of them seem to make things of the same kind.

The situation is clearer if one looks with care the [GR framework documentation](http://gr-framework.org/julia.html), where the three modules `GR`, `gr3` and `jlgr` are separated. But perhaps it would help to avoid confusion if the APIs are also separated. I think that from the development perspective it would also be cleaner.

For instance, many of the high-level functions defined in the module `jlgr` have some kind of "shadow" of themselves in that module; e.g. `plot` is defined there, but the module also imports `GR`, so `jlgr` also knows about `GR.plot`, which is the same as `jlgr.plot` because `GR` includes `jlgr.jl` and defines its own `plot` function equal to `jlgr.plot`. Now, after `using GR` the user has all the functions from the different modules mixed, and it is not easy to distinguish them. (I have re-read this paragraph and I acknowledge that it is really twisted, but in my opinion this is only a reflection of the actual complication of the code.)

Since the main module of this package is `GR`, I assume that its core functionality is the low level functions, which are the basis for higher level APIs like `jlgr` inside the package itself, or other packages like [Plots](https://juliaplots.github.io/). If this is the case, I'd suggest to remove the line that includes `jlgr.jl` in the module `GR` and the redefinition (and exportation) of the high-level functions in that module. It would be cleaner - and in my opinion easier to understand - if those high-level functions were exported directly by `jlgr`.

The only small drawback from the user perspective is that in order to write `plot(x,y)` and get the plot on your screen, you should be `using GR.jlgr`, instead of the simpler `using GR`. But if it is properly explained in the main page, I don't think that it is making things much more complicated, and it might help to make some things easier to understand.

**Comment by heliosdrm** [comment]
By the way, the name of the module `jlgr` does not follow [Julia's naming conventions](https://docs.julialang.org/en/v1/manual/style-guide/index.html#Use-naming-conventions-consistent-with-Julia-base/-1). I hope I'm not crossing the line if I also suggest to rename it with a capitalized and more meaningful name, like `GRUtils` or something like that.

**Comment by jheinen** [comment]
I'd say that'd be a good choice. Alternatively we could rename the core part to `GRCore` and leave the old name (`GR`) for the convenience functions.

**Comment by heliosdrm** [comment]
I have some ideas about refactoring `jlgr`, which would address this issue and also introduce the following improvements:

* Solve some open issues like saving multiple subplots (#222), or the behavior of `hold` in different environments (#206).
* Use parametric types, and reduce the amount of arrays and collections of abstract type.
* Replace large `if-else` blocks by functions with type dispatch.
* (Hopefully) improve performance, thanks to the two previous points.
* Use meta-programming to reduce repetition of code.
* In general, make the code simpler and easier to extend and maintain.

The API would be essentially the same as the one provided by `jlgr`, but I could make this in a separate module `GRUtils` that would not break anything of what is already working. If the result is good enough, `jlgr` might eventually be dropped without much trouble.

However, since @jheinen has other plans for the high-level functionality of GR based on C rather than on Julia (#222), I'm not sure about the worthiness of this proposal. I would be happy to work on it, if the term for those plans is unclear, and such a new Julia module may improve the experience of using and developing GR for a reasonable time. But if the new high-level interface will come early, I would save the effort. :smile:

**Comment by heliosdrm** [comment]
I eventually decided to try and make that experiment.

It is here, as announced in discourse a couple of weeks ago (although it is much better documented right now):

https://github.com/heliosdrm/GRUtils.jl

It only lacks the "serialized" and "meta" features of `jlgr`. If the concept is compatible with the philosophy of GR, I'd be happy to bring come from there. Otherwise, it can be left as yet another front end for GR.

**Comment by heliosdrm** [comment]
Some context information that may help to decide if [GRUtils](https://github.com/heliosdrm/GRUtils.jl) might be eventually merged into GR, replacing part of jlgr.

In this moment, GRUtils has nearly all the functionalities provided by jlgr, with the following notes:

* I have not yet released the functions that need GR>0.41 support, i.e. polar and nonuniform heatmaps. I'm waiting for GR 0.42 or higher to be registered.
* The interface of some functions is slightly different, [as explained in the documentation](https://heliosdrm.github.io/GRUtils.jl/stable/#Relationship-with-GR's-API-1).
* The "meta"- and "serialized" stuff is completely left out.

The data structures used by GRUtils are different from those made by jlgr, but I have tried to make it so that they are translatable into each other -- at least for the "standard" plots. Although the structures used by GRUtils are explained in the documentation, here I provide a description centred on the relationships and differences with jlgr:

## plots in jlgr

In jlgr all plots are made through a global object called `plt`, which is of type `PlotObject`. That type contains:

* `plt.args`: a vector whose items are tuples with four arrays and a string. The arrays are normally referred to as `x`, `y`, `z` and `c`, and contain the coordinates of the geometries that are plotted. The string is normally used for the format specification of line plots and similar.
* `kvs`: a dictionary with additional data that are used to define the characteristics of the plot. The most important item is probably `plt.kvs[:kind]`, whose value is a symbol that describes the kind of the plot (`:line`, `:scatter`, `:contour`, etc.).

* `obj`: another dictionary that is used as a bridge with copies of the data contained in `plt.kvs`, during the creation and modification of the plot.

## plots in GRUtils

In GRUtils there is also a type `PlotObject` that is used to contain those data, although structured in a different way.

* The structure of the tuples contained in `plt.args` is replicated in (part of) the type `Geometry`, which also contains slots for `x`, `y`, `z` and `c` vectors, as well as for a string `spec` -- among other things. One of the fields of the type `PlotObject` is `geoms`, a vector of such objects, which therefore is roughy equivalent to `plt.args`.
* The data gathered in `plt.kvs` is distributed in the fields of other elements contained in every `PlotObject`: the viewport, axes, legend, colorbar, and also a dictionary of `attributes`.
* The kind of the plot, i.e. the way in which the information contained in the `PlotObject` should be used to draw it, is not encoded in a single item like `plt.kvs[:kind]`. There is a `kind` for the plot axes, and also a `kind` for *each* geometry contained in `geoms`. `PlotObject` items also have a field called `kind`, which might be used to facilitate the "translation" into the language of `jlgr`, although `GRUtils` does not use it.

## compatibilities and incompatibilities

Thus, in theory I expect that all plots made by `jlgr` could be translatable into the structure of `GRUtils`. On the other hand, the usual combinations of geometry and axes kinds should be also translatable to the structure of `jlgr`, although there may be more complex combinations that do not fit.

In particular, it may be noted that the `Geometry` type also has its own dictionary of attributes, so you can combine geometries with different values for the same attribute in the same plot.

Another big difference is that GRUtils has a type `Figure` to collect different `PlotObjects`. The global object used in the creation and modification of plots is such a `Figure`, whereas in jlgr you only work with a global `PlotObject`. (You can still see previous plots on the device if you use `subplot`, until you wipe them out; but albeit visible, the older plots are unreachable once you create a new one.) 

## next steps ?

This being said, the possibility of translating the simpler plots between the "languages" of jlgr and GRUtils should be sufficient to fill the main gap left by GRUtils, which is the absence of `send_meta` and `send_serialized`, seemingly bound to the structure defined by `jlgr`. Those utilities might be moved to a specific module of `GR`, and I could create the necessary "translation" functions.

To complete the integration of GRUtils in GR (if eventually accepted to do it), it would suffice to include the files of the repository. I would also suggest to remove all the aliases given in `GR.jl` for the functions of jlgr, as explained in the first post of this issue.

There are various open issues of GR that are already solved with the structure of GRUtils. The most important one I think is the one of working with multiple figures, which I think cannot be addressed without big changes in jlgr. But there are also other minor issues, e.g. the recent issue of transparency when using `savefig` (#262), `imshow` not working with files (#239) and others. And I have also found it easy to incorporate new plot kinds like multiple plot bars, error bars or quiver plots (although I'm of course not objective to judge that, since I made the code).

**Comment by jheinen** [comment]
Many thanks for the great work on GRUtils. I will take a look at your package and hopefully find a way to merge the two modules. Nevertheless, I would like to first fix the bugs in `jlgr` - GRUtils will be a big help - and then continue the development on a common path. Would that be acceptable to you?

**Comment by heliosdrm** [comment]
That's great. While I'm improving GRUtils with new features, I'll also continue submitting PRs to GR for those bugs that I find the way to fix without breaking the current API -- as I have being doing during this time.
